consensus/istanbul/core/roundstate.go

```diff
@@ -182,30 +220,34 @@ func (s *roundStateImpl) Round() *big.Int {
 	return s.round
 }
 
-func (s *roundStateImpl) changeRound(nextRound *big.Int, validatorSet istanbul.ValidatorSet) {
+func (s *roundStateImpl) changeRound(nextRound *big.Int, validatorSet istanbul.ValidatorSet, nextProposer istanbul.Validator) {
 	s.state = StateAcceptRequest
 	s.round = nextRound
 	s.desiredRound = nextRound
 
+	// TODO MC use old valset
 	s.prepares = newMessageSet(validatorSet)
 	s.commits = newMessageSet(validatorSet)
+	s.proposer = nextProposer
 
 	// ??
 	s.preprepare = nil
 }
 
-func (s *roundStateImpl) StartNewRound(nextRound *big.Int, validatorSet istanbul.ValidatorSet) {
+func (s *roundStateImpl) StartNewRound(nextRound *big.Int, validatorSet istanbul.ValidatorSet, nextProposer istanbul.Validator) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
-	s.changeRound(nextRound, validatorSet)
+	s.changeRound(nextRound, validatorSet, nextProposer)
 }
 
-func (s *roundStateImpl) StartNewSequence(view *istanbul.View, validatorSet istanbul.ValidatorSet, parentCommits MessageSet) {
+func (s *roundStateImpl) StartNewSequence(view *istanbul.View, validatorSet istanbul.ValidatorSet, nextProposer istanbul.Validator, parentCommits MessageSet) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 
-	s.changeRound(view.Round, validatorSet)
+	s.validatorSet = validatorSet
+
+	s.changeRound(view.Round, validatorSet, nextProposer)
```

Comment:
Should `view.Round` always be 0 here?  If so, perhaps checking for that here and log and error and return an error if it's not.

Reply:
It is always 0, as you can see here: https://github.com/celo-org/celo-blockchain/blob/83b6a40c2526803d64d75db4e8e1d02938453104/consensus/istanbul/core/core.go#L335

Again, the problem is that `startNewRound` is a very confusing function, that should be refactor. But even if i was tempted, i felt it wouldn't get me much love from my reviewers... so i left that for a later PR ;)

Reply:
As we discussed over slack, the plan is to modify `StartNewSequence` to accept a sequence number instead of a view number, and to modify `resetRoundState` to check that the round number is zero here:  https://github.com/celo-org/celo-blockchain/pull/628/files/83b6a40c2526803d64d75db4e8e1d02938453104#diff-7d85e6de59ad4e84300834cf03eb41d1R400

Reply:
done